### PR TITLE
modules: littlefs: move CMake glue code from littlefs repo into Zephyr

### LINKED ISSUE #40559: Move LittlefFS configuration header and CMakeLists.txt from module to zephyr/modules
**Is your enhancement proposal related to a problem? Please describe.**
Currently we modify one of littlefs header files, lfs_util.h, to allow it work with Zephyr.
The LittleFS, as noted here https://github.com/zephyrproject-rtos/zephyr/pull/40484#issuecomment-975353232 by @Laczen , can be build with configuration header file given as "-DDLFS_CONFIG=" option to compiler which seems to be better option, and allows to have nearly unmodified (with except to addition zephyr/modyle.yml file) LittleFS fork.

**Describe the solution you'd like**
1) move CMakeLists.txt out of modules/fs/littlefs to zephyr/module/littlefs
2) copy the modified version of lfs_util.h to zephyr/module/littlefs/lfs_zephyr_config.h and remove unneeded code from it;
3) modify the CMakeLists.txt for the LittleFS module to include the "-DDLFS_CONFIG=lfs_zephyr_config.h;
4) revert modifications to module/fs/littlefs/lfs_util.h and other Zephyr specific changes, except for addition of zephyr/module.yml, from LittleFS module.

**Describe alternatives you've considered**
Leaving this at it is

**Additional context**
The subject has been briefly discussed here: https://github.com/zephyrproject-rtos/zephyr/pull/40484

### PR DESCRIPTION
**Pull request by tejlmand**
Fixes: #40559

Moving CMakeLists.txt gluecode into Zephyr repo minimizes the patching
needed in LittleFS repo.

It further allow Zephyr to generate its own LFS config header based upon
Kconfig settings which further reduces the patching needed in LittleFS.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

### CODE REVIEW

**Comment by tejlmand** [comment]
@poojamg FYI

**Comment by tejlmand** [comment]
@nashif scancode for license check fails because we have two files originating from LittleFS copied into Zephyr repo to minimize the patching needed in the LittleFS.

One is the `lfs_util.h` that is copied and modified because it is the configuration template for LittleFS.
Second is the lfs_util.c containing a basic crc function that is disabled when using a custom lfs config header.

Those two files are under BSD license, and as I only make changes two them we can't change the license, but to mu knowledge a BSD licensed file should be acceptable in Zephyr.

**Comment by poojamg** [comment]
This changes are ok to me, now how do I enable LFS_READONLY feature of lfs_util.h?

**Comment by tejlmand** [comment]
> This changes are ok to me, now how do I enable LFS_READONLY feature of lfs_util.h?

I made a comment on that here: https://github.com/zephyrproject-rtos/littlefs/pull/8#issuecomment-1010282171

**Comment by poojamg** [comment]
> > This changes are ok to me, now how do I enable LFS_READONLY feature of lfs_util.h?
> 
> I made a comment on that here: [zephyrproject-rtos/littlefs#8 (comment)](https://github.com/zephyrproject-rtos/littlefs/pull/8#issuecomment-1010282171)

Thanks this will work for me! will add the kconfig changes ..

File: modules/littlefs/CMakeLists.txt
Comment (de-nordic) [inline modules/littlefs/CMakeLists.txt:17]: The PR looks so spotless that this is the only place I can stick to: indentation shorter than in lines 8-11 that are at the same indentation level.

File: modules/littlefs/CMakeLists.txt
Comment (de-nordic) [inline modules/littlefs/CMakeLists.txt:11]: Closing bracket is indented differently than in line 18

Comment (tejlmand) [reply modules/littlefs/CMakeLists.txt]: Fixed.

Danger of simply moving code unaltered from one repo to another and then make extra additions :stuck_out_tongue:

File: modules/littlefs/CMakeLists.txt
Comment (tejlmand) [reply modules/littlefs/CMakeLists.txt]: Fixed.
Code removed.

As we always adds `-imacro autoconf.h` we can avoid the `configure_file()` stage and have config header directly.

**Review by de-nordic — APPROVED** [review]
Looks OK, waiting for sha change after the LitteFS update gets merged.

**Comment by dkalowsk** [comment]
Is it possible to get this resolved in time for 3.1 inclusion?  @poojamg and I let this one drop off the radar unfortunately for 3.0 which is when we were hoping to get it merged in.  If I am understanding this correctly, the license scan is the only blocking bit?

File: subsys/fs/CMakeLists.txt
Comment (SebastianBoe) [inline subsys/fs/CMakeLists.txt:16]: Isn't it cleaner to link with LITTLEFS ?

File: modules/littlefs/CMakeLists.txt
Comment (SebastianBoe) [inline modules/littlefs/CMakeLists.txt:8]: Why is this necessary?

Someone that links with LITTLEFS is not already linking with zephyr_interface ?

Which library is that ?

File: modules/littlefs/zephyr_lfs_crc.c
Comment (SebastianBoe) [inline modules/littlefs/zephyr_lfs_crc.c:1]: The file is in the zephyr littlefs module, so the zephyr_lfs_ prefix is just noise IMHO.

Should rename to crc.c

File: modules/littlefs/zephyr_lfs_config.h
Comment (SebastianBoe) [inline modules/littlefs/zephyr_lfs_config.h:8]: It's not possible to #include lfs_util.h instead of copying it is it ?

File: modules/littlefs/zephyr_lfs_crc.c
Comment (SebastianBoe) [inline modules/littlefs/zephyr_lfs_crc.c:31]: I don't understand the need to copy this file.

Why don't we just add https://github.com/littlefs-project/littlefs/blob/master/lfs_util.c to the build ?

Comment (SebastianBoe) [reply modules/littlefs/zephyr_lfs_crc.c:31]: Oh, I see we are:

https://github.com/zephyrproject-rtos/zephyr/pull/41729/files#diff-01ac9904de8007b215faf5b86a7ece70d162e5a70244a85a26b2166bb4d8e95bR14

so what's the problem? The define ifdef's out the function?

Then maybe we can undefine it, or have a source file that #include's lfs_util.c.

It's unfortunate to copy crypto code like this due to it being difficult to track security bugs when the code is copied.

**Review by SebastianBoe — DISMISSED** [review]
changes requested

File: modules/littlefs/zephyr_lfs_crc.c
Comment (SebastianBoe) [reply modules/littlefs/zephyr_lfs_crc.c:31]: Maybe littlefs will listen to reason so we don't have to deal with these hacks.

https://github.com/littlefs-project/littlefs/pull/647

Comment (Laczen) [reply modules/littlefs/zephyr_lfs_crc.c:31]: @SebastianBoe, I think the idea is to replace this with a call to zephyr's builtin crc32. It is just not done at the moment.

**Comment by de-nordic** [comment]
@tejlmand Any response to @SebastianBoe reqests?

File: modules/littlefs/zephyr_lfs_config.h
Comment (nvlsianpu) [reply modules/littlefs/zephyr_lfs_config.h:8]: Litttlefs Project stands in its lfs_util.h: 
> // Users can override lfs_util.h with their own configuration by defining
// LFS_CONFIG as a header file to include (-DLFS_CONFIG=lfs_config.h).

so this is OK

File: modules/littlefs/zephyr_lfs_crc.c
Comment (tejlmand) [reply modules/littlefs/zephyr_lfs_crc.c:31]: > @SebastianBoe, I think the idea is to replace this with a call to zephyr's builtin crc32. It is just not done at the moment.

Exactly, as the commit message states:
https://github.com/zephyrproject-rtos/zephyr/pull/41729/commits/399c1b88e2c794a62b7602fa76edfd0635dddb26
> This allows to ensure identical behavior when moving glue code into
Zephyr repo, and allow us to change to an existing crc implementation
provided natively by Zephyr at a later time.

The crc implementation provided with LFS is very naive, as it even says directly in the code.
https://github.com/zephyrproject-rtos/zephyr/blob/399c1b88e2c794a62b7602fa76edfd0635dddb26/modules/littlefs/zephyr_lfs_crc.c#L9-L10

In my opinion this CRC **is not** suitable for production code.

Therefore we should switch to using a Zephyr CRC implementation as this issue describes:
https://github.com/zephyrproject-rtos/zephyr/issues/43369

However such work is outside the scope of this PR, as this PR intends to move glue code out of LittleFS and thus reduce LittleFS patching and make it easier to upmerge.

Comment (tejlmand) [reply modules/littlefs/zephyr_lfs_crc.c:31]: also note, with this approach, when #43369 is completed, then this commit can easily be reverted and thus removing this file again which is the end goal.
https://github.com/zephyrproject-rtos/zephyr/pull/41729/commits/399c1b88e2c794a62b7602fa76edfd0635dddb26

File: modules/littlefs/CMakeLists.txt
Comment (tejlmand) [reply modules/littlefs/CMakeLists.txt:8]: note sure, code is moved from here: https://github.com/zephyrproject-rtos/littlefs/blob/zephyr/CMakeLists.txt

and I do not know if there is a reason, so I prefer not to change.

Changing this is outside the scope of moving gluecode into Zephyr and should be cleaned up independently.

**Review by Laczen — APPROVED** [review]
Looks good, thanks.

**Review by carlescufi — APPROVED** [review]
This is a special case of [external contributions](https://docs.zephyrproject.org/latest/contribute/external.html), The source code in question was already present in the project, as a module, and it is now being moved to the main tree. Given this, we do not need to go through the full process as [described in the doc](https://docs.zephyrproject.org/latest/contribute/external.html), and we can instead merge this as-is and convert those two files to become Zephyr-owned automatically.

**Pull request merged**
